src/EventStore.Core/DataStructures/ProbabilisticFilter/MemoryMappedFileBloomFilter/MemoryMappedFileBloomFilter.cs

```diff
@@ -0,0 +1,279 @@
+		/*
+		    Bloom filter implementation based on the following paper by Adam Kirsch and Michael Mitzenmacher:
+		    "Less Hashing, Same Performance: Building a Better Bloom Filter"
+		    https://www.eecs.harvard.edu/~michaelm/postscripts/rsa2008.pdf
+
+		    Only two 32-bit hash functions can be used to simulate additional hash functions of the form g(x) = h1(x) + i*h2(x)
+
+		    First cache line (64 bytes) contains the header and nothing else
+		    Subsequent cachelines contain 60 bytes of bloom filter bits then a 4 byte hash of those bits.
+		*/
+		protected static readonly ILogger Log = Serilog.Log.ForContext<MemoryMappedFileBloomFilter>();
+		public const long MinSizeKB = 10;
+		public const long MaxSizeKB = 4_000_000;
+		public const double RecommendedFalsePositiveProbability = 0.02;
+		private const int CacheLineSize = BloomFilterIntegrity.CacheLineSize;
+		private const int NumHashFunctions = 6;
+		private readonly long _numBits;
+		private readonly Header _header;
+		private readonly byte* _pointer = null;
+
+		private readonly IHasher[] _hashers = {
+			new XXHashUnsafe(seed: 0xC0015EEDU),
+			new Murmur3AUnsafe(seed: 0xC0015EEDU),
+			new XXHashUnsafe(seed: 0x13375EEDU),
+			new Murmur3AUnsafe(seed: 0x13375EEDU)
+		};
+
+		private readonly FileStream _fileStream;
+		private readonly long _fileSize;
+		private readonly MemoryMappedFile _mmf;
+		private readonly MemoryMappedViewAccessor _mmfWriteAccessor;
+		private readonly object _writeLock = new();
+
+		private bool _disposed;
+
+		public int CorruptionRebuildCount => _header.CorruptionRebuildCount;
+
+		/// <summary>
+		/// Bloom filter implementation which uses a memory-mapped file for persistence.
+		/// </summary>
+		/// <param name="path">Path to the bloom filter file</param>
+		/// <param name="size">Size of the bloom filter in bytes</param>
+		///
+		/// Add and MightContain can be called concurrently on different threads.
+		/// If the call to Add is complete then a call to MightContain will find the added value.
+		public MemoryMappedFileBloomFilter(string path, bool create, long size, int corruptionRebuildCount = 0) {
+			Ensure.NotNull(path, nameof(path));
+
+			if (size < MinSizeKB * 1000 || size > MaxSizeKB * 1000) {
+				throw new ArgumentOutOfRangeException(nameof(size), $"size should be between {MinSizeKB:N0} and {MaxSizeKB:N0} KB inclusive");
+			}
+
+			_numBits = size * 8;
+
+			_fileStream = new FileStream(
+				path,
+				// todo: OpenOrCreate can be CreateNew if the mininode closed the file before declaring itself 'stopped'
+				create ? FileMode.OpenOrCreate : FileMode.Open,
+				FileAccess.ReadWrite,
+				FileShare.ReadWrite);
+
+			if (Header.Size > CacheLineSize)
+				throw new Exception("header needs to fit in a cache line");
+
+			// add room for filter bits, rounded up to nearest cacheline
+			_fileSize = GetBytePositionInFile(_numBits) / CacheLineSize * CacheLineSize;
```

Comment:
It appears that there are two off-by-ones:
- Should GetBytePositionInFile(_numBits) be GetBytePositionInFile(_numBits - 1) ? (since it's a 0-based index and the largest bit is _numBits - 1)
-  / CacheLineSize * CacheLineSize is taking the floor, so there'll be one cache line missing if (_numBits-1) is not an exact multiple of CacheLineSize (I'm assuming that the line below: `_fileSize += CacheLineSize` is for the header)

Reply:
for the second point, i think that I'm wrong - the header's cache line is being taken into consideration inside GetBytePositionInFile() and the line `_fileSize += CacheLineSize` is actually for the extra cache line if it's not an exact multiple, right?

Reply:
Good questions, in reverse order:

- `GetBytePositionInFile` accounts for the header (thats what im trying to indicate with the 'infile' part of the name), so the added cacheline on the next line is inded to account for the fact that we floored it in the calculation, not to make room for the header
- `GetBytePositionInFile(_numBits - 1)` would get the us the position of the last byte in the file that holds filter bits. But to calculate the length needed we want the first byte that doesnt hold filter bits. e.g. if the position of the last byte is 123 then the length is 124 (before rounding up to the nearest cacheline)

Reply:
@timothycoleman thanks for all the clarifications!

regarding `_numBits` vs `_numBits - 1`, i still have some doubts:
take the case where the last filter bit (`_numBit - 1`) lies at a 60-byte boundary just before the 4-byte hash, then the next bit (`_numBit`) will be in the next cache line. With the current code after rounding up, there'll be an extra 64-byte cache line in the file (it should not cause any harm though)

Reply:
thats true, say _numbits is 60*8=480 then `GetBytePositionInFile(480)` will return 128. which is correct

however since its already exactly a cache line we neededn't add on an extra one (but as you say it doesnt harm if we do so)

Reply:
@timothycoleman ok, this extra cache line doesn't happen if we use _numBits - 1 since it can be read as:
- Find the cache line to which the last filter bit belongs too
- Round down to the nearest cache line: we're now at the second-to-last cache line
- Add an extra cache line

But since it doesn't cause any harm, i'm marking it as resolved :-)